### PR TITLE
Bugfix/move and reference use

### PR DESCRIPTION
**Pull request by GlynLeine**

### CODE REVIEW

**Comment by GlynLeine** [comment]
multiguard PR should probably be merged first!

File: args/core/containers/sparse_map.hpp [hunk 1]
```diff
@@ -420,14 +421,15 @@ namespace args::core
...
+				k = *itr_key;
+				m_sparse[k] = m_size;
 				++m_size;
 			}
+			else
+				k = key;
 
-
-			return m_dense_value[m_sparse[key]];
+			return m_dense_value[m_sparse[k]];
```
Comment (Algo-ryth-mix) [inline args/core/containers/sparse_map.hpp]: mind explaining a bit what is going on here ?

File: args/core/async/readonly_rw_spinlock.hpp [hunk 2]
```diff
@@ -161,26 +395,137 @@ namespace args::core::async
...
-				lock.readState.store(readonly_rw_spinlock::read, std::memory_order_release);
+				locked = true;
+				for (uint i : unlockedLocks)
+				{
+					if (m_locks[i]->try_lock(write))
+						unlockedLocks.erase(i);
+					else
+						locked = false;
+				}
+			} while (!locked);
```
Comment (Algo-ryth-mix) [inline args/core/async/readonly_rw_spinlock.hpp]: how does this actually prevent deadlocks ? shouldn't all locks be unlocked on an failed attempt in order to not spin endlessly ?

**Review by Algo-ryth-mix — CHANGES_REQUESTED** [review]

File: args/core/containers/sparse_map.hpp (hunk 1 above)
Comment (Algo-ryth-mix) [reply args/core/containers/sparse_map.hpp]: (in the form of comments preferably)

Comment (GlynLeine) [reply args/core/containers/sparse_map.hpp]: uhm sure, but it's quite standard. using a move operation might invalidate the previous rvalue or lvalue ref you had. thus you cannot use that item anymore.

File: args/core/async/readonly_rw_spinlock.hpp (hunk 2 above)
Comment (GlynLeine) [reply args/core/async/readonly_rw_spinlock.hpp]: yes you're right...

File: args/core/containers/sparse_map.hpp (hunk 1 above)
Comment (Algo-ryth-mix) [reply args/core/containers/sparse_map.hpp]: yes, but being quite standard does not mean it is particularly readable as a whole
```cpp
		value_reference operator[](key_type&& key)
		{
			key_type k;
			if (!contains(key))
			{
				if (m_size >= m_capacity)
					reserve(m_size + 1);

				auto itr_value = m_dense_value.begin() + m_size;
				*itr_value = std::forward<value_type>(value_type());

				auto itr_key = m_dense_key.begin() + m_size;
				*itr_key = std::move(key);
				k = *itr_key;
				m_sparse[k] = m_size;
				++m_size;
			}
			else
				k = key;

			return m_dense_value[m_sparse[k]];
		}
```
here is the piece of code in question without github diff stuff around it. Not super readbale imo.
It does not help that we have itr_key, key, and k here which are seamingly the same things (and all quite terrible names for variables)

File: args/core/async/readonly_rw_spinlock.hpp [hunk 3]
```diff
@@ -324,35 +324,51 @@ namespace args::core::async
...
 	public:
 		/**@brief Creates readonly multi-guard and locks for Read-only.
 		 */
 		template<typename lock_type1 = readonly_rw_spinlock, typename lock_type2 = readonly_rw_spinlock, typename... lock_typesN>
 		readonly_multiguard(lock_type1& lock1, lock_type2& lock2, lock_typesN&... locks)
 		{
-			m_locks.push_back(&lock1);
-			m_locks.push_back(&lock2);
-			(m_locks.push_back(&locks), ...);
+			fill<sizeof...(locks) + 2>(lock1, lock2, locks...);
```
Comment (Algo-ryth-mix) [inline args/core/async/readonly_rw_spinlock.hpp]: why even use fill here ?
```suggestion
			m_locks = {{ &lock1, &lock2, &locks... }};
```
longer example with better optimization as well https://godbolt.org/z/rWx7vx

File: args/core/containers/sparse_map.hpp (hunk 1 above)
Comment (GlynLeine) [reply args/core/containers/sparse_map.hpp]: Untitled

is this okay?

File: args/core/ecs/entityquery.cpp [hunk 4]
```diff
@@ -25,7 +25,7 @@ namespace args::core::ecs
 	{
 		hashed_sparse_set<id_type> componentTypes;
 		if (m_id)
-			componentTypes = m_registry.getComponentTypes(m_id);
+			componentTypes = m_registry.getComponentTypes(m_id); /// comment this glyn.
```
Comment (Algo-ryth-mix) [inline args/core/ecs/entityquery.cpp:28]: I heard that talking to yourself is not healthy, I am unsure about writing yourself.

**Review by Algo-ryth-mix — APPROVED** [review]

File: args/core/containers/sparse_map.hpp (hunk 1 above)
Comment (Algo-ryth-mix) [reply args/core/containers/sparse_map.hpp]: yes

File: args/core/async/readonly_rw_spinlock.hpp (hunk 3 above)
Comment (GlynLeine) [reply args/core/async/readonly_rw_spinlock.hpp]: got it

**Pull request merged**